### PR TITLE
fix(hmr): Prevent retrieving properties in undefined object in jsHMR

### PR DESCRIPTION
**Pull request by shirohana**
Hello,

I met a strange error when working with HMR and `react-hot-loader`, here's the screenshot:

截圖 2020-06-14 下午6 21 36

maybe related to https://github.com/webpack/webpack-dev-server/issues/2231.

---

I've searched almost the possibility across `react-hot-loader`, `webpack-dev-middleware` and `webpack-hot-middleware`, or even I used Koa in the wrong way. I excluded most of the customization, finally I guess this is an implicit bug when mixing `webpack@^5` and `react-hot-loader`.

I skipped creating an issue but a pull request, hope that won't let you feel bad (ノω・。)

### Minimal reproducible project with step-by-step:

Also more informations.

https://github.com/shirohana/webpack-react-hmr-undefined

### Notable versions

```
"babel-loader": "^8.1.0"
"react-hot-loader": "^4.12.21"
"webpack": "^5.0.0-beta.17"
"webpack-dev-middleware": "^4.0.0-rc.1"
"webpack-hot-middleware": "^2.25.0"
```

---

**What kind of change does this PR introduce?**

> A tiny quickfix to avoid some unexpected but common use cases.

**Did you add tests for your changes?**

> No, no new test needed.

**Does this PR introduce a breaking change?**

> No.

**What needs to be documented once your changes are merged?**

> No need to be documented, just a tiny patch.

### CODE REVIEW

**Comment by shirohana** [comment]
Hi there,

I'm sorry about this pull request... seems this problem should be solved in `react-hot-loader` not HMR of `webpack`.

After cross testing, the main problem is from `react-hot-loader` definitely, not sure is about module cache invalidating or require blocking chain.

Hope that I can do some meaningful contributes and join webpack community in some day, sorry for annoying and thank you for bringing amazing project (´・ω・｀ʃ♡ƪ)

File: lib/hmr/JavascriptHotModuleReplacement.runtime.js [hunk 1]
```diff
@@ -210,6 +210,7 @@ module.exports = function () {
 				// removed self-accepted modules should not be required
 				appliedUpdate[outdatedModuleId] !== warnUnexpectedRequire &&
 				// when called invalidate self-accepting is not possible
+				$moduleCache$[moduleId] &&
 				!$moduleCache$[moduleId].hot._selfInvalidated
```
Comment (sokra) [inline lib/hmr/JavascriptHotModuleReplacement.runtime.js:214]:
``` suggestion
				!$moduleCache$[outdatedModuleId].hot._selfInvalidated
```

**Comment by sokra** [comment]
Do you want to fix the new problem?

For creating a test case try to look into `test/hotCases`

File: lib/hmr/JavascriptHotModuleReplacement.runtime.js (hunk 1 above)
Comment (shirohana) [reply lib/hmr/JavascriptHotModuleReplacement.runtime.js:214]: Oh (´⊙ω⊙`)！
I'll patch that.

**Comment by shirohana** [comment]
@sokra Thank you for the advice, I'll read the guide and try to create some tests.

**Comment by shirohana** [comment]
Hi there,

I finally got the most minimal case which can fire the error, and only the first commit applied, the test will passed.

I'm not sure about that is it I named the test case correctly btw (´・ω・｀)

**Comment by sokra** [comment]
Thanks

**Pull request merged**